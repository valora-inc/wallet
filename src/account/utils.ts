import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
