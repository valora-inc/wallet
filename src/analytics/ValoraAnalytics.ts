import { createClient, SegmentClient } from '@segment/analytics-react-native'
import LegacySegment from '@segment/analytics-react-native-legacy'
import { AdjustPlugin } from '@segment/analytics-react-native-plugin-adjust'
import { ClevertapPlugin } from '@segment/analytics-react-native-plugin-clevertap'
import { FirebasePlugin } from '@segment/analytics-react-native-plugin-firebase'
import { sha256FromString } from 'ethereumjs-util'
import _ from 'lodash'
import { Platform } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import { check, PERMISSIONS, request, RESULTS } from 'react-native-permissions'
import { AsyncStoragePersistor } from 'src/analytics/AsyncStoragePersistor'
import { AppEvents } from 'src/analytics/Events'
import { InjectTraits } from 'src/analytics/InjectTraits'
import { AnalyticsPropertiesList } from 'src/analytics/Properties'
import { getCurrentUserTraits } from 'src/analytics/selectors'
import {
  DEFAULT_TESTNET,
  E2E_TEST_STATSIG_ID,
  FIREBASE_ENABLED,
  isE2EEnv,
  SEGMENT_API_KEY,
  STATSIG_API_KEY,
  STATSIG_ENV,
} from 'src/config'
import { store } from 'src/redux/store'
import { getDefaultStatsigUser } from 'src/statsig'
import Logger from 'src/utils/Logger'
import { Statsig } from 'statsig-react-native'

const TAG = 'ValoraAnalytics'

interface DeviceInfoType {
  AppName: string
  Brand: string
  BuildNumber: string
  BundleId: string
  Carrier: string
  DeviceId: string // this is the device model + version
  FirstInstallTime: number
  FontScale: number
  FreeDiskStorage: number
  InstallReferrer: string
  InstanceID: string
  LastUpdateTime: number
  Manufacturer: string
  MaxMemory: number
  Model: string
  ReadableVersion: string
  SystemName: string
  SystemVersion: string
  TotalDiskCapacity: number
  TotalMemory: number
  UniqueID: string // this is the unique id of the device, which maps to deviceId in the data
  UserAgent: string
  Version: string
  isEmulator: boolean
  isTablet: boolean
  UsedMemory: number
}

async function getDeviceInfo(): Promise<DeviceInfoType> {
  return {
    AppName: DeviceInfo.getApplicationName(),
    Brand: DeviceInfo.getBrand(),
    BuildNumber: DeviceInfo.getBuildNumber(),
    BundleId: DeviceInfo.getBundleId(),
    Carrier: await DeviceInfo.getCarrier(),
    DeviceId: DeviceInfo.getDeviceId(),
    FirstInstallTime: await DeviceInfo.getFirstInstallTime(),
    FontScale: await DeviceInfo.getFontScale(),
    FreeDiskStorage: await DeviceInfo.getFreeDiskStorage(),
    InstallReferrer: await DeviceInfo.getInstallReferrer(),
    InstanceID: await DeviceInfo.getInstanceId(),
    LastUpdateTime: await DeviceInfo.getLastUpdateTime(),
    Manufacturer: await DeviceInfo.getManufacturer(),
    MaxMemory: await DeviceInfo.getMaxMemory(),
    Model: DeviceInfo.getModel(),
    ReadableVersion: DeviceInfo.getReadableVersion(),
    SystemName: DeviceInfo.getSystemName(),
    SystemVersion: DeviceInfo.getSystemVersion(),
    TotalDiskCapacity: await DeviceInfo.getTotalDiskCapacity(),
    TotalMemory: await DeviceInfo.getTotalMemory(),
    UniqueID: await DeviceInfo.getUniqueId(),
    UserAgent: await DeviceInfo.getUserAgent(),
    Version: DeviceInfo.getVersion(),
    isEmulator: await DeviceInfo.isEmulator(),
    isTablet: DeviceInfo.isTablet(),
    UsedMemory: await DeviceInfo.getUsedMemory(),
  }
}

class ValoraAnalytics {
  sessionId: string = ''
  deviceInfo: DeviceInfoType | undefined

  private currentScreenId: string | undefined
  private prevScreenId: string | undefined
  private segmentClient: SegmentClient | undefined

  async init() {
    let uniqueID
    try {
      if (!SEGMENT_API_KEY) {
        throw Error('API Key not present, likely due to environment. Skipping enabling')
      }
      this.segmentClient = createClient({
        debug: __DEV__,
        trackAppLifecycleEvents: true,
        trackDeepLinks: true,
        writeKey: SEGMENT_API_KEY,
        storePersistor: AsyncStoragePersistor,
      })

      // TODO: Remove this code once there is a break in onboarding experiments (note: removing this code will change all users anonymousId's)
      // This is a temporary workaround for a side effect of upgrading to Segment v2
      // Segment v2 generates a different anonymousId for the user than Segment v1
      // In order to avoid disrupting ongoing onboarding experiments by changing the anonymousId mid-experiment we use the legacy Segment client to get the anonymousId
      await LegacySegment.setup('dummy-key', {}) // not necessary to provide a valid key, but is necessary to call setup.
      const anonymousId = await LegacySegment.getAnonymousId()
      await this.segmentClient.userInfo.set({ anonymousId })

      this.segmentClient.add({ plugin: new InjectTraits() })
      this.segmentClient.add({ plugin: new AdjustPlugin() })
      this.segmentClient.add({ plugin: new ClevertapPlugin() })
      if (FIREBASE_ENABLED) {
        this.segmentClient.add({ plugin: new FirebasePlugin() })
      }

      try {
        const deviceInfo = await getDeviceInfo()
        this.deviceInfo = deviceInfo
        uniqueID = deviceInfo.UniqueID
        this.sessionId = sha256FromString('0x' + uniqueID.split('-').join('') + String(Date.now()))
          .toString('hex')
          .slice(2)
      } catch (error) {
        Logger.error(TAG, 'getDeviceInfo error', error)
      }

      Logger.info(TAG, 'Segment Analytics Integration initialized!')
    } catch (error) {
      Logger.error(TAG, `Segment setup error: ${error.message}\n`, error)
    }

    try {
      const statsigUser = getDefaultStatsigUser()
      // getAnonymousId causes the e2e tests to fail
      let overrideStableID: string = E2E_TEST_STATSIG_ID
      if (!isE2EEnv) {
        if (!this.segmentClient) {
          throw new Error('segmentClient is undefined, cannot get anonymous ID')
        }
        overrideStableID = this.segmentClient.userInfo.get().anonymousId
      }
      Logger.debug(TAG, 'Statsig stable ID', overrideStableID)
      await Statsig.initialize(STATSIG_API_KEY, statsigUser, {
        // StableID should match Segment anonymousId
        overrideStableID,
        environment: STATSIG_ENV,
        localMode: isE2EEnv,
      })
    } catch (error) {
      Logger.warn(TAG, `Statsig setup error`, error)
    }
  }

  isEnabled() {
    // Remove __DEV__ here to test analytics in dev builds
    return !__DEV__ && store.getState().app.analyticsEnabled
  }

  startSession(
    eventName: typeof AppEvents.app_launched,
    eventProperties: AnalyticsPropertiesList[AppEvents.app_launched]
  ) {
    this.track(eventName, {
      deviceInfo: this.deviceInfo,
      ...eventProperties,
    })

    this.requestTrackingPermissionIfNeeded().catch((error) => {
      Logger.error(TAG, 'Failure while requesting tracking permission', error)
    })
  }

  getSessionId() {
    return this.sessionId
  }

  track<EventName extends keyof AnalyticsPropertiesList>(
    ...args: undefined extends AnalyticsPropertiesList[EventName]
      ? [EventName] | [EventName, AnalyticsPropertiesList[EventName]]
      : [EventName, AnalyticsPropertiesList[EventName]]
  ) {
    const [eventName, eventProperties] = args

    if (!this.isEnabled()) {
      Logger.debug(TAG, `Analytics is disabled, not tracking event ${eventName}`)
      return
    }

    if (!this.segmentClient) {
      Logger.debug(TAG, `segmentClient undefined, not tracking event ${eventName}`)
      return
    }

    const props: {} = {
      ...this.getSuperProps(),
      ...eventProperties,
    }

    Logger.info(TAG, `Tracking event ${eventName} with properties:`, props)

    this.segmentClient.track(eventName, props).catch((err) => {
      Logger.error(TAG, `Failed to track event ${eventName}`, err)
    })
  }

  identify(userID: string | null, traits: {}) {
    if (!this.isEnabled()) {
      Logger.debug(TAG, `Analytics is disabled, not tracking user ${userID}`)
      return
    }

    // Only identify user if userID (walletAddress) is set
    if (!userID) {
      return
    }

    if (!this.segmentClient) {
      Logger.debug(TAG, `segmentClient is undefined, not tracking user ${userID}`)
      return
    }
    // The firebase segment plugin can't handle null or undefined values
    const safeTraits = _.omitBy(traits, _.isNil)

    this.segmentClient.identify(userID, safeTraits).catch((err) => {
      Logger.error(TAG, `Failed to identify user ${userID}`, err)
      throw err
    })
  }

  page(screenId: string, eventProperties = {}) {
    if (!this.isEnabled()) {
      Logger.debug(TAG, `Analytics is disabled, not tracking screen ${screenId}`)
      return
    }

    if (!this.segmentClient) {
      Logger.debug(TAG, `segmentClient is undefined, not tracking screen ${screenId}`)
      return
    }

    if (screenId !== this.currentScreenId) {
      this.prevScreenId = this.currentScreenId
      this.currentScreenId = screenId
    }

    const props: {} = {
      ...this.getSuperProps(),
      ...eventProperties,
    }

    this.segmentClient.screen(screenId, props).catch((err) => {
      Logger.error(TAG, 'Error tracking page', err)
    })
  }

  async reset() {
    if (!this.segmentClient) {
      Logger.debug(TAG, `segmentClient is undefined, not resetting`)
      return
    }
    try {
      await this.segmentClient.flush()
      await this.segmentClient.reset()
    } catch (error) {
      Logger.error(TAG, 'Error resetting analytics', error)
    }
  }

  private async requestTrackingPermissionIfNeeded() {
    // TODO: remove `isE2EEnv` and set permission via Detox when we upgrade
    if (Platform.OS !== 'ios' || isE2EEnv) {
      return
    }

    const appTrackingPermission = await check(PERMISSIONS.IOS.APP_TRACKING_TRANSPARENCY)
    Logger.debug(TAG, `iOS tracking permission: ${appTrackingPermission}`)
    if (appTrackingPermission !== RESULTS.DENIED) {
      // The permission has already been requested / is not requestable
      // See https://github.com/zoontek/react-native-permissions#permissions-statuses
      return
    }

    Logger.debug(TAG, `iOS requesting tracking permission`)
    this.track(AppEvents.request_tracking_permission_started, {
      currentPermission: appTrackingPermission,
    })
    const newAppTrackingPermission = await request(PERMISSIONS.IOS.APP_TRACKING_TRANSPARENCY)
    Logger.debug(TAG, `iOS tracking permission after request: ${newAppTrackingPermission}`)
    if (newAppTrackingPermission === RESULTS.GRANTED) {
      this.track(AppEvents.request_tracking_permission_accepted, {
        newPermission: newAppTrackingPermission,
      })
    } else {
      this.track(AppEvents.request_tracking_permission_declined, {
        newPermission: newAppTrackingPermission,
      })
    }
  }

  // Super props, i.e. props sent with all events
  private getSuperProps() {
    const traits = getCurrentUserTraits(store.getState())
    // Prefix super props with `s` so they don't clash with events props
    const prefixedSuperProps = Object.fromEntries(
      Object.entries({
        ...traits,
        currentScreenId: this.currentScreenId,
        prevScreenId: this.prevScreenId,
      }).map(([key, value]) => [`s${key.charAt(0).toUpperCase() + key.slice(1)}`, value])
    )

    return {
      // Legacy super props
      timestamp: Date.now(),
      sessionId: this.sessionId,
      userAddress: traits.walletAddress,
      celoNetwork: DEFAULT_TESTNET,
      // Prefixed super props
      ...prefixedSuperProps,
      // Statsig prop, won't be read properly by Statsig if prefixed
      statsigEnvironment: STATSIG_ENV,
    }
  }
}

let isInitialized = false
type KeysOfType<T, TProp> = { [P in keyof T]: T[P] extends TProp ? P : never }[keyof T]
type ValoraAnalyticsKeyFunction = KeysOfType<ValoraAnalytics, Function>
// Type checked function keys to queue until `init` has finished
const funcsToQueue = new Set<ValoraAnalyticsKeyFunction>([
  'startSession',
  'track',
  'identify',
  'page',
])
let queuedCalls: Function[] = []

function isFuncToQueue(prop: string | number | symbol): prop is ValoraAnalyticsKeyFunction {
  return funcsToQueue.has(prop as ValoraAnalyticsKeyFunction)
}

/**
 * Use a proxy to queue specific calls until async `init` has finished
 * So all events are sent with the right props to our initialized analytics integrations
 */
export default new Proxy(new ValoraAnalytics(), {
  get: function (target, prop, receiver) {
    if (!isInitialized) {
      if (prop === 'init') {
        return new Proxy(target[prop], {
          apply: (target, thisArg, argumentsList) => {
            return Reflect.apply(target, thisArg, argumentsList).finally(() => {
              isInitialized = true
              // Init finished, we can now process queued calls
              for (const fn of queuedCalls) {
                fn()
              }
              queuedCalls = []
            })
          },
        })
      } else if (isFuncToQueue(prop)) {
        return new Proxy(target[prop], {
          apply: (target, thisArg, argumentsList) => {
            queuedCalls.push(() => Reflect.apply(target, thisArg, argumentsList))
            Logger.debug(TAG, `Queued call to ${prop}`, ...argumentsList)
          },
        })
      }
    }
    return Reflect.get(target, prop, receiver)
  },
})
