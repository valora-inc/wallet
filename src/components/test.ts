function groupNumber(value: string) {
  return value.replace(/\B(?=(\d{3})+(?!\d))(?<!\.\d*)/g, 'group')
}

function actualTest(value: string, decimalSeparator: string, groupingSeparator: string) {
  return groupNumber(value).replaceAll('.', decimalSeparator).replaceAll('group', groupingSeparator)
}

console.log(actualTest('1.234.567,9999', ',', '.')) // 1,234,567.9999
