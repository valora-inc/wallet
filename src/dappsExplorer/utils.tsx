import { DappV2WithCategoryNames } from 'src/dapps/types'

/***
 * Calculates a score based on the search term and dapp
 * - +2 if the name matches the search term in a case sensitive way
 * - +1 if the name matches the search term in a case insensitive way
 * - +1 if the description matches the search term in a case insensitive way
 * - +1 if the category matches the search term in a case insensitive way
 * @param dapp: dapp with categoryNames
 * @param searchTerm: the user input search string
 * @returns {number} the score of the dapp 0 - 5
 */
export const calculateSearchScore = (dapp: DappV2WithCategoryNames, searchTerm: string) => {
  if (searchTerm === '') return 0
  const nameMatchingCaseScore = dapp.name.includes(searchTerm) ? 2 : 0
  const nameScore = dapp.name.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const descriptionScore = dapp.description.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const categoryScore = dapp.categoryNames.some((category) =>
    category.toLowerCase().includes(searchTerm.toLowerCase())
  )
    ? 1
    : 0
  return nameMatchingCaseScore + nameScore + descriptionScore + categoryScore
}
