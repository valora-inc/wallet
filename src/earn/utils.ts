import BigNumber from 'bignumber.js'
import { getFeatureGate } from 'src/statsig'
import { StatsigFeatureGates } from 'src/statsig/types'
import { SwapTransaction } from 'src/swap/types'
import { Network, NetworkId } from 'src/transactions/types'
import { INTERNAL_RPC_SUPPORTED_NETWORKS } from 'src/viem'
import { networkIdToNetwork } from 'src/web3/networkConfig'

/**
 * Helper function to check if gas is subsidized for the given networkId
 * @param networkId - networkId to check
 * @returns boolean
 */
export function isGasSubsidizedForNetwork(networkId: NetworkId) {
  return (
    (INTERNAL_RPC_SUPPORTED_NETWORKS as readonly Network[]).includes(
      networkIdToNetwork[networkId]
    ) && getFeatureGate(StatsigFeatureGates.SUBSIDIZE_STABLECOIN_EARN_GAS_FEES)
  )
}

/**
 * Gets the swap to amount in decimals given a swap transaction and the swap
 * from amount in decimals. This could also be taken from the buy amount, but
 * that is in wei and we would need the token decimals. This follows what is
 * done in the SwapScreen
 * @param object.swapTransaction - swap transaction object
 * @param object.fromAmount - swap from amount in decimals (usually from the user input)
 * @returns swap to amount in decimals
 */
export function getSwapToAmountInDecimals({
  swapTransaction,
  fromAmount,
}: {
  swapTransaction: SwapTransaction
  fromAmount: BigNumber
}) {
  return fromAmount.multipliedBy(swapTransaction.price)
}
