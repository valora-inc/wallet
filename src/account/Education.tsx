import React, { useEffect, useRef, useState } from 'react'
import {
  Image,
  ImageSourcePropType,
  NativeScrollEvent,
  ScrollView,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
} from 'react-native'
import { NativeSafeAreaViewProps, SafeAreaView } from 'react-native-safe-area-context'
import { OnboardingEvents } from 'src/analytics/Events'
import { ScrollDirection } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import Button, { BtnTypes } from 'src/components/Button'
import Pagination from 'src/components/Pagination'
import BackChevron from 'src/icons/BackChevron'
import Times from 'src/icons/Times'
import DrawerTopBar from 'src/navigator/DrawerTopBar'
import { navigateBack } from 'src/navigator/NavigationService'
import { TopBarIconButton } from 'src/navigator/TopBarButton'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import progressDots from 'src/styles/progressDots'
import variables from 'src/styles/variables'

export enum EmbeddedNavBar {
  Close = 'Close',
  Drawer = 'Drawer',
}

export enum EducationTopic {
  onboarding = 'onboarding',
  backup = 'backup',
  celo = 'celo',
}

interface EducationStep {
  image: ImageSourcePropType | null
  topic: EducationTopic
  title: string
  // If set to true, title is displayed at the top
  isTopTitle?: boolean
  text?: string
}

export type Props = NativeSafeAreaViewProps & {
  embeddedNavBar: EmbeddedNavBar | null
  stepInfo: EducationStep[]
  buttonType?: BtnTypes
  buttonText: string
  finalButtonType?: BtnTypes
  finalButtonText: string
  dotStyle?: StyleProp<ViewStyle>
  activeDotStyle?: StyleProp<ViewStyle>
  onFinish: () => void
}

const Education = (props: Props) => {
  const {
    style,
    embeddedNavBar,
    stepInfo,
    buttonType = BtnTypes.SECONDARY,
    buttonText,
    finalButtonType = BtnTypes.PRIMARY,
    finalButtonText,
    dotStyle = progressDots.circlePassive,
    activeDotStyle = progressDots.circleActive,
    onFinish,
    ...passThroughProps
  } = props

  const [currentIndex, setCurrentIndex] = useState(0)
  // This variable tracks the last scrolled to carousel screen, so that impression
  // events are not dispatched twice for the same carousel screen
  const lastViewedIndex = useRef(-1)
  // Scroll View Ref for button clicks
  const scrollViewRef = useRef<ScrollView>(null)

  const handleScroll = (event: { nativeEvent: NativeScrollEvent }) => {
    const { topic } = stepInfo[currentIndex]
    const nextIndex = Math.round(event.nativeEvent.contentOffset.x / variables.width)
    if (nextIndex === currentIndex) {
      return
    }

    const direction = nextIndex > currentIndex ? ScrollDirection.next : ScrollDirection.previous
    if (topic === EducationTopic.backup) {
      ValoraAnalytics.track(OnboardingEvents.backup_education_scroll, {
        currentStep: currentIndex,
        direction: direction,
      })
    } else if (topic === EducationTopic.celo) {
      ValoraAnalytics.track(OnboardingEvents.celo_education_scroll, {
        currentStep: currentIndex,
        direction: direction,
      })
    } else if (topic === EducationTopic.onboarding) {
      ValoraAnalytics.track(OnboardingEvents.onboarding_education_scroll, {
        currentStep: currentIndex,
        direction: direction,
      })
    }

    setCurrentIndex(Math.round(event.nativeEvent.contentOffset.x / variables.width))
  }

  useEffect(() => {
    const { topic } = stepInfo[currentIndex]
    if (stepInfo.length > 0 && lastViewedIndex.current < currentIndex) {
      lastViewedIndex.current = currentIndex
    }
    if (topic === EducationTopic.onboarding) {
      ValoraAnalytics.track(OnboardingEvents.onboarding_education_step_impression, {
        step: currentIndex,
      })
    }
  }, [currentIndex])

  if (!stepInfo.length) {
    // No Steps, no slider
    return null
  }

  const goBack = () => {
    const { topic } = stepInfo[currentIndex]
    if (currentIndex === 0) {
      if (topic === EducationTopic.backup) {
        ValoraAnalytics.track(OnboardingEvents.backup_education_cancel)
      } else if (topic === EducationTopic.celo) {
        ValoraAnalytics.track(OnboardingEvents.celo_education_cancel)
      }
      navigateBack()
    } else {
      scrollViewRef.current?.scrollTo({ x: variables.width * (currentIndex - 1), animated: true })
    }
  }

  const nextStep = () => {
    // If we are on the last step, call the onFinish function otherwise scroll to the next step
    currentIndex === stepInfo.length - 1
      ? onFinish()
      : scrollViewRef.current?.scrollTo({ x: variables.width * (currentIndex + 1), animated: true })
  }

  const renderEmbeddedNavBar = () => {
    switch (embeddedNavBar) {
      case EmbeddedNavBar.Close:
        return (
          <View style={styles.top} testID="Education/top">
            <TopBarIconButton
              testID={`Education/${currentIndex === 0 ? 'Close' : 'Back'}Icon`}
              onPress={goBack}
              icon={currentIndex === 0 ? <Times /> : <BackChevron color={colors.black} />}
            />
          </View>
        )
      case EmbeddedNavBar.Drawer:
        return <DrawerTopBar testID="DrawerTopBar" />
      default:
        return null
    }
  }

  return (
    <SafeAreaView testID="Education" style={[styles.root, style]} {...passThroughProps}>
      {renderEmbeddedNavBar()}
      <View style={styles.container}>
        <ScrollView
          ref={scrollViewRef}
          horizontal={true}
          pagingEnabled={true}
          showsHorizontalScrollIndicator={false}
          onScroll={handleScroll}
          testID="ScrollContainer"
        >
          {stepInfo.map((step: EducationStep, i: number) => (
            <ScrollView
              showsVerticalScrollIndicator={false}
              contentContainerStyle={styles.contentContainer}
              style={styles.swipedContent}
              key={i}
            >
              {step.isTopTitle && <Text style={styles.headingTop}>{step.title}</Text>}
              <View style={styles.swipedContentInner}>
                {step.image && (
                  <Image source={step.image} style={styles.bodyImage} resizeMode="contain" />
                )}
                {!step.isTopTitle && <Text style={styles.heading}>{step.title}</Text>}
                {!!step.text && <Text style={styles.bodyText}>{step.text}</Text>}
              </View>
            </ScrollView>
          ))}
        </ScrollView>
        <Pagination
          style={styles.pagination}
          count={stepInfo.length}
          activeIndex={currentIndex}
          dotStyle={dotStyle}
          activeDotStyle={activeDotStyle}
        />
        <Button
          testID="Education/progressButton"
          onPress={nextStep}
          text={currentIndex === stepInfo.length - 1 ? finalButtonText : buttonText}
          type={currentIndex === stepInfo.length - 1 ? finalButtonType : buttonType}
          touchableStyle={
            currentIndex === stepInfo.length - 1 ? { backgroundColor: '#EEB93C' } : undefined
          }
        />
      </View>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  root: {
    flex: 1,
  },
  contentContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flexGrow: 1,
    paddingHorizontal: 12,
  },
  container: {
    flex: 1,
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
    paddingBottom: 24,
  },
  heading: {
    marginTop: 24,
    ...fontStyles.h2,
    textAlign: 'center',
  },
  headingTop: {
    ...fontStyles.h1,
    marginTop: 26,
    alignSelf: 'flex-start',
  },
  bodyText: {
    ...fontStyles.regular,
    textAlign: 'center',
    paddingTop: 16,
    marginBottom: 24,
  },
  bodyImage: {
    alignSelf: 'center',
    marginBottom: 24,
  },
  swipedContent: {
    width: variables.width - 2 * variables.contentPadding,
    margin: variables.contentPadding,
  },
  swipedContentInner: {
    flex: 1,
    justifyContent: 'center',
  },
  top: {
    paddingLeft: 24,
    paddingVertical: 16,
    flexDirection: 'row',
    width: '100%',
  },
  pagination: {
    paddingBottom: variables.contentPadding,
  },
})

export default Education
