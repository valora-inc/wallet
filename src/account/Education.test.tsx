import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { navigateBack } from 'src/navigator/NavigationService'
import { createMockStore } from 'test/utils'
import Education, { EducationTopic, EmbeddedNavBar } from '../../src/account/Education'

jest.mock('src/analytics/ValoraAnalytics')

const BUTTON_TEXT = 'Done'

const educationProps = {
  stepInfo: [
    {
      image: null,
      topic: EducationTopic.celo,
      title: 'Step 1',
      text: 'The Journey Begins',
    },
  ],
  buttonText: 'next',
  embeddedNavBar: EmbeddedNavBar.Close,
  finalButtonText: BUTTON_TEXT,
  onFinish: jest.fn(),
}

describe('Education', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={createMockStore({})}>
        <Education {...educationProps} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('finishes when pressed', () => {
    const edu = render(<Education {...educationProps} />)
    fireEvent.press(edu.UNSAFE_getByProps({ text: BUTTON_TEXT }))
    expect(educationProps.onFinish).toBeCalled()
  })

  it('navigates back', () => {
    const edu = render(<Education {...educationProps} />)
    fireEvent.press(edu.getByTestId('Education/CloseIcon'))
    expect(navigateBack).toBeCalled()
  })

  it('onboarding step info analytics event fires', () => {
    // Onboarding analytics event
    const onboardingProps = { ...educationProps }
    onboardingProps.stepInfo[0].topic = EducationTopic.onboarding
    render(<Education {...educationProps} />)
    expect(ValoraAnalytics.track).toHaveBeenCalledWith(
      OnboardingEvents.onboarding_education_step_impression,
      {
        step: 0,
      }
    )
  })
})
