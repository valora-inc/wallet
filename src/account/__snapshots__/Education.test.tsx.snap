// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Education renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
  testID="Education"
>
  <View
    style={
      {
        "flexDirection": "row",
        "paddingLeft": 24,
        "paddingVertical": 16,
        "width": "100%",
      }
    }
    testID="Education/top"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      nativeBackgroundAndroid={
        {
          "attribute": "selectableItemBackgroundBorderless",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="Education/CloseIcon"
    >
      <svg
        height={16}
        style={{}}
        testID="Times"
        viewBox="0 0 16 16"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.9999 2.00146L1.99994 14.0015"
          stroke="#2E3338"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          style={{}}
        />
        <path
          d="M1.99994 2.00146L13.9999 14.0015"
          stroke="#2E3338"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          style={{}}
        />
      </svg>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexWrap": "wrap",
        "justifyContent": "center",
        "paddingBottom": 24,
      }
    }
  >
    <RCTScrollView
      horizontal={true}
      onScroll={[Function]}
      pagingEnabled={true}
      showsHorizontalScrollIndicator={false}
      testID="ScrollContainer"
    >
      <View>
        <RCTScrollView
          contentContainerStyle={
            {
              "alignItems": "center",
              "flexGrow": 1,
              "justifyContent": "center",
              "paddingHorizontal": 12,
            }
          }
          showsVerticalScrollIndicator={false}
          style={
            {
              "margin": 16,
              "width": 718,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#2E3338",
                    "fontFamily": "Jost-Medium",
                    "fontSize": 22,
                    "lineHeight": 28,
                    "marginTop": 24,
                    "textAlign": "center",
                  }
                }
              >
                Step 1
              </Text>
              <Text
                style={
                  {
                    "color": "#2E3338",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "marginBottom": 24,
                    "paddingTop": 16,
                    "textAlign": "center",
                  }
                }
              >
                The Journey Begins
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RCTScrollView>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 100,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "#1AB775",
                "flexDirection": "row",
                "height": 48,
                "justifyContent": "center",
                "minWidth": 120,
                "opacity": undefined,
                "paddingHorizontal": 24,
                "paddingVertical": 5,
              },
              undefined,
            ]
          }
          testID="Education/progressButton"
        >
          <Text
            maxFontSizeMultiplier={1}
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
                "marginLeft": 0,
                "marginRight": 0,
              }
            }
          >
            Done
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
